spec/services/slack_service_spec.rb

```diff
@@ -0,0 +1,236 @@
+      expect(data.name).to eql("cap")
+      expect(data.real_name).to eql("Steve Rogers")
+      expect(data.first_name).to eql("Steve")
+      expect(data.email).to eql("steve@ombulabs.com")
+      expect(data.tz).to eql("America/New_York")
+    end
+
+    it "failure: given a missing user - returns an error" do
+      result, data = SlackService.find_slack_user("D00DD0DDD0D", SlackServiceSpec::SlackClientMock)
+      expect(result).to eql(:error)
+      expect(data).to eql("User Not Found")
+    end
+  end
+
+  describe ".send_message" do
+    it "success: given a user id - sends a simple message" do
+      result, data = SlackService.send_message("A00AA1AAA", "A message", SlackServiceSpec::SlackClientMock)
+      expect(result).to eql(:ok)
+      expect(data).to eql("Message Sent")
+    end
+
+    it "success: given a user id - sends a complex message" do
+      result, data = SlackService.send_message(
+        "A00AA1AAA",
+        {
+          text: "A message",
+          attachments: [{"pretext": "pre-hello", "text": "text-world"}],
+          blocks: [{"type": "section", "text": {"type": "plain_text", "text": "Hello world"}}]
+        },
+        SlackServiceSpec::SlackClientMock
+      )
+      expect(result).to eql(:ok)
+      expect(data).to eql("Message Sent")
+
+      # The following asserts send_message receives th
+    end
+
+    it "failure: given a missing user - returns an error" do
+      result, data = SlackService.send_message("D00DD0DDD0D", "A message", SlackServiceSpec::SlackClientMock)
+      expect(result).to eql(:error)
+      expect(data).to eql("Slack Channel Not Found")
+    end
+  end
+
+  describe ".cobble_message_params" do
+    it "success: receives simple text messages" do
+      result = SlackService.cobble_message_params("The ID", "The Message")
+      expect(result).to eql({channel: "The ID", text: "The Message"})
+    end
+
+    it "success: receives complex text messages" do
+      result = SlackService.cobble_message_params("The ID", {
+        text: "A message",
+        attachments: [{"pretext": "pre-hello", "text": "text-world"}],
+        blocks: [{"type": "section", "text": {"type": "plain_text", "text": "Hello world"}}]
+      })
+      expect(result).to eql({
+        channel: "The ID",
+        text: "A message",
+        attachments: [{"pretext": "pre-hello", "text": "text-world"}],
+        blocks: [{"type": "section", "text": {"type": "plain_text", "text": "Hello world"}}]
+      })
+    end
+
+    it "failure: requires one of text, attachments, or blocks" do
+      expect { SlackService.cobble_message_params("The ID", {}) }.to raise_error(SlackService::MessageFormatError, "Message missing")
+    end
+
+    it "failure: rejects aditional keys" do
+      expect { SlackService.cobble_message_params("The ID", {text: "The Message", other: "other"}) }.to raise_error(SlackService::MessageFormatError, "Message format incorrect")
+    end
+  end
+end
+
+module SlackServiceSpec
```

Comment:
I think we could mock this module using a  test `double` https://relishapp.com/rspec/rspec-mocks/docs/basics/test-doubles instead of overwriting the module

Reply:
So my usual approach is that if I have a dependency I can't control or sandbox I inject it and set up a detailed Mock of what I'm expecting in the tests.  Every method on `SlackService` receives a "Client" as it's final value. Outside of test callers are expected to pass `Slack::Web::Client` from the Slack gem... in tests I build these mock classes so I can display what I think the API is doing.

That said - I don't have a strong opinion on that (not in Ruby anyway). If we drop these for simple Mocks I'd recommend me removing the dependency injection from `SlackService`.